### PR TITLE
rbac: introduce a general model to abstract the RBAC policy

### PR DESCRIPTION
**Pull request by yangminzhu**
The model decouples the policy parsing and envoy filter config generation which makes it much easier to support different implementation. It also makes it easier in unit test and policy simulation.

Note, there is only trivial changes to the current unit tests, this is intentionally to make sure the refactor doesn't break anything.

https://github.com/istio/istio/issues/13489 https://github.com/istio/istio/issues/11569

### CODE REVIEW

File: pilot/pkg/networking/plugin/authz/conversion.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
+package authz
```
Comment (pitlv2109) [inline pilot/pkg/networking/plugin/authz/conversion.go]: Add Istio Apache License template. Same with some other files.

File: pilot/pkg/networking/plugin/authz/conversion.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	envoy_rbac "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2alpha"
+	envoy_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher"
+
+	"istio.io/istio/pilot/pkg/networking/plugin/authz/matcher"
+	"istio.io/istio/pilot/pkg/networking/plugin/authz/rbacfilter"
+	authn_v1alpha1 "istio.io/istio/pilot/pkg/security/authn/v1alpha1"
+	"istio.io/istio/pkg/spiffe"
+)
+
+func permissionForKeyValues(key string, values []string) *envoy_rbac.Permission {
```
Comment (pitlv2109) [inline pilot/pkg/networking/plugin/authz/conversion.go]: I know that it is a refactoring, but maybe we can add some docstring here for this function? Same with some other big functions.

File: pilot/pkg/networking/plugin/authz/validate.go [hunk 3]
```diff
@@ -1,77 +0,0 @@
-// Copyright 2018 Istio Authors
```
Comment (pitlv2109) [inline pilot/pkg/networking/plugin/authz/validate.go]: Why are these checking deleted?

File: pilot/pkg/networking/plugin/authz/conversion.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	return pg.OrPermissions()
+}
+
+// principalForKeyValues converts an Istio first class field (e.g. namespaces, groups, ips, as opposed
+// to properties fields) to Envoy RBAC config and returns said field as a Principal or returns nil
+// if the field is empty.
+func principalForKeyValues(key string, values []string, forTCPFilter bool) *envoy_rbac.Principal {
+	pg := rbacfilter.PrincipalGenerator{}
+	for _, value := range values {
+		id := principalForKeyValue(key, value, forTCPFilter)
```
Comment (pitlv2109) [inline pilot/pkg/networking/plugin/authz/conversion.go]: nit: maybe there is a better name than `id`? Same with other places that are using `principalForKeyValues`.

File: pilot/pkg/networking/plugin/authz/conversion.go (hunk 2 above)
Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/conversion.go]: Done

File: pilot/pkg/networking/plugin/authz/conversion.go (hunk 1 above)
Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/conversion.go]: Done

File: pilot/pkg/networking/plugin/authz/validate.go (hunk 3 above)
Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/validate.go]: It's moved to `model.go`/`model_test.go`

File: pilot/pkg/networking/plugin/authz/conversion.go (hunk 4 above)
Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/conversion.go]: Renamed to `principal`.

**Pull request merged**